pandas/_typing.py

```diff
@@ -11,12 +11,13 @@
 from pandas.core.dtypes.generic import (
     ABCExtensionArray, ABCIndexClass, ABCSeries, ABCSparseSeries)
 
-AnyArrayLike = Union[ABCExtensionArray,
-                     ABCIndexClass,
-                     ABCSeries,
-                     ABCSparseSeries,
-                     np.ndarray]
-ArrayLike = Union[ABCExtensionArray, np.ndarray]
+AnyArrayLike = TypeVar('AnyArrayLike',
```

Comment:
Have to think through impact of this some more. After looking at this my thought was we might need to mark this as covariant, but it doesn't look like that is supported in generic functions (which was one of the motivations here):

https://www.python.org/dev/peps/pep-0484/#covariance-and-contravariance

If you have any thoughts yourself please let me know!

Reply:
It's quite a read, I will look at it later today.

Reply:
I don't think we need convariants here. Even after reading it a couple times I still don't get it very well, but it looks are it has something to do with sub/super class, which doesn't looks like the case here.

Reply:
OK. I think this is passing without covariant because we might only be using in generic functions for now. I think it would be a problem when parametrizing containers since these by default are invariant (i.e. we could parametrize a container with ABCIndexClass but not ABCIndex) but can deal with that when that happens

Quoting relevant part of PEP 484:

> By default generic types are considered invariant in all type variables, which means that values for variables annotated with types like List[Employee] must exactly match the type annotation -- no subclasses or superclasses of the type parameter (in this example Employee) are allowed.

Reply:
The end solution for parametrization may be to add a `AnyArrayLike_co` since that's a convention suggested by the PEP anyway

Reply:
OK, Remaining types are:

    DatetimeLikeScalar = Type[Union[Period, Timestamp, Timedelta]]
    Dtype = Union[str, np.dtype, ExtensionDtype]
    FilePathOrBuffer = Union[str, Path, IO[AnyStr]]

- Union in `DatetimeLikeScalar` is nested don't know if it should be converted to `TypeVar`
- Converting `Dtype` to TypeVar starts giving errors.